File: LeetCode/2053. Kth Distinct String in an Array/2053.cpp
class Solution {
public:
    string kthDistinct(vector<string>& arr, int k) {
        // Step 1: Count occurrences of each string
        unordered_map<string, int> countMap = countOccurrences(arr);

        // Step 2: Collect distinct strings
        list<string> distinctStrings = collectDistinctStrings(arr, countMap);

        // Step 3: Return the k-th distinct string
        return getKthDistinctString(distinctStrings, k);
    }

private:
    unordered_map<string, int> countOccurrences(const vector<string>& arr) {
        unordered_map<string, int> countMap;
        for (const string& str : arr) {
            countMap[str]++;
        }
        return countMap;
    }

    list<string> collectDistinctStrings(const vector<string>& arr, const unordered_map<string, int>& countMap) {
        list<string> distinctStrings;
        for (const string& str : arr) {
            if (countMap.at(str) == 1) {
                distinctStrings.push_back(str);
            }
        }
        return distinctStrings;
    }

    string getKthDistinctString(const list<string>& distinctStrings, int k) {
        if (k > distinctStrings.size()) {
            return "";
        }
        auto it = distinctStrings.begin();
        advance(it, k - 1);
        return *it;
    }
};

/*
class Solution {
public:
    string kthDistinct(vector<string>& arr, int k) {
        unordered_map<string, int> mp;
        string ans = "";

        for(string x : arr) {
            mp[x]++;
        }

        for(auto& s : arr) {
            if(mp[s] == 1 && --k == 0)
                return s;
        }
 
        return "";
    }
};*/

File: LeetCode/2053. Kth Distinct String in an Array/2053.java
class Solution {
    public String kthDistinct(String[] arr, int k) {
        // Step 1: Count occurrences of each string
        Map<String, Integer> countMap = countOccurrences(arr);

        // Step 2: Collect distinct strings
        List<String> distinctStrings = collectDistinctStrings(arr, countMap);

        // Step 3: Return the k-th distinct string
        return getKthDistinctString(distinctStrings, k);
    }

    private Map<String, Integer> countOccurrences(String[] arr) {
        Map<String, Integer> countMap = new HashMap<>();
        for (String str : arr) {
            countMap.put(str, countMap.getOrDefault(str, 0) + 1);
        }
        return countMap;
    }

    private List<String> collectDistinctStrings(String[] arr, Map<String, Integer> countMap) {
        List<String> distinctStrings = new ArrayList<>();
        for (String str : arr) {
            if (countMap.get(str) == 1) {
                distinctStrings.add(str);
            }
        }
        return distinctStrings;
    }

    private String getKthDistinctString(List<String> distinctStrings, int k) {
        return (k <= distinctStrings.size()) ? distinctStrings.get(k - 1) : "";
    }
}

/*
 * class Solution {
 * public String kthDistinct(String[] arr, int k) {
 * List<String> uniqueValInArray = new LinkedList<>();
 * int totalDistinctVal = 0;
 * Set<String> nonDistinctSet = new HashSet<>();
 * Set<String> distinctSet = new HashSet<>();
 * for(String s:arr) {
 * if(!nonDistinctSet.contains(s)) {
 * if(!distinctSet.contains(s)) {
 * distinctSet.add(s);
 * totalDistinctVal++;
 * } else {
 * nonDistinctSet.add(s);
 * distinctSet.remove(s);
 * totalDistinctVal--;
 * }
 * }
 * }
 * if(totalDistinctVal<k) {
 * return "";
 * }
 * for(String s:arr) {
 * if(distinctSet.contains(s)){
 * k--;
 * if(k==0){
 * return s;
 * }
 * }
 * }
 * return "";
 * }
 * }
 */